tests/devices/test_caching.py

```diff
@@ -0,0 +1,277 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests for caching executions of the quantum device."""
+import numpy as np
+import pytest
+
+import pennylane as qml
+from pennylane.tape.measure import expval
+from pennylane.tape import qnode, QNode
+from pennylane.devices import DefaultQubit
```

Comment:
I noticed a lot of the tests check whether `DefaultQubit` has its `apply` method called. What about testing `Device`?

Reply:
I just had a go now and it seems like it will complicate the tests a bit - `QubitDevice` has abstract methods such as `apply()` which we have to define. In `test_qubit_device` a mock device is defined, but it uses a simpler `apply()` method: `lambda self, x: None`. For these tests, we do care that different QNode inputs map to different circuits (e.g., so we can see that same inputs use the cache, while other inputs don't).

I think it's ok to use `DefaultQubit`, since it inherits directly from `QubitDevice` and does not alter the `execute()` method.